src/object/obj_internal.h

```diff
@@ -44,6 +44,13 @@
  */
 #define IO_BYPASS_ENV	"DAOS_IO_BYPASS"
 
+/* EC parity is stored in a private address range that is selected by setting
+ * the most-significant bit of the offset (an unsigned long). This effectively
+ * limits the addressing of user extents to the lower 63 bits of the offset
+ * range. The client stack should enforce this limitation.
+ */
+#define PARITY_INDICATOR (1UL << 63)
```

Comment:
I'm considering another way for the EC parity handling.
That firstly scan the IOD/RECXs (combine continuous RECX when needed), and remember which set of IOD/RECXs need EC encode. Then allocate needed buffer for parity, encode it, and in another structure to store the index in original IOD/RECXs and new parity code buffer. Transfer the parity (bulk) and the index to server. i.e. don't mix the parity code together with original data so need not change/insert to original IOD/sgl.
By this way can avoid the REALLOC/copy IOD/sgl in your DAOS-2203 branch, the handling seems clearer. Need not this PARITY_INDICATOR flag bit also.
But I think we probably can go with your current implementation first and get your code in your draft branch in master ASAP, and refine that later if we all think that's better.

Reply:
Right now, the structures used for the update are part of the API. This is why parity is mixed with user data (if we use the existing structures, then the IOD, SGL are the only place to put the parity). The parity indicator bit is needed, even if we change how parity is transferred in the update.. The extents have to be uniquely addressed under a given dkey/akey, and without setting this bit (and reserving it for EC), there will be collisions between user data and system-generated parity.

Reply:
ok, understand, thanks. I mean another potential option is alloc/manage the parity code buffer in another separate place/structure (with parity buffer and index point to original IOD/recx). The IOD/sgl structure is part of API, while how to manage the EC parity buffer is internal to daos client/server RPC protocol. It seems to me that mixing parity code together with user data need some careful logic to manipulate the IOD/sgl struct as in your branch. Separate them possibly is simpler, but I think we can go with your current code first.